sys/linux/init_images.go

```diff
@@ -0,0 +1,243 @@
+	if zr.left == 0 {
+		return 0, io.EOF
+	}
+	toRead := zr.left
+	if toRead > len(p) {
+		toRead = len(p)
+	}
+	for i := 0; i < toRead; i++ {
+		p[i] = 0
+	}
+	zr.left -= toRead
+	return toRead, nil
+}
+
+const imageMaxSize = 129 << 20
+
+func fixUpImageSegments(parsed *mountImageArgs) {
+	parsed.filterSegments(func(i int, segment *mountImageSegment) bool {
+		if segment.parseError != nil {
+			return false
+		}
+		return segment.offset.Val < imageMaxSize && segment.size.Val < imageMaxSize
+	})
+	// Overwriting of the image multiple times is not efficient and complicates image extraction in Go.
+	// So let's make segments non-overlapping.
+	sort.Stable(&sortSegments{parsed: parsed})
+	resizeSegment := func(s *mountImageSegment, size uint64) {
+		s.size.Val = size
+		s.data.SetData(s.data.Data()[0:size])
+	}
+
+	newSize := parsed.size.Val
+	for idx, segment := range parsed.segments {
+		actualSize := uint64(len(segment.data.Data()))
+		if segment.size.Val != actualSize {
+			segment.size.Val = actualSize
+		}
+		if idx > 0 {
+			// Adjust the end of the previous segment.
+			prevSegment := parsed.segments[idx-1]
+			if prevSegment.offset.Val+prevSegment.size.Val > segment.offset.Val {
+				resizeSegment(prevSegment, segment.offset.Val-prevSegment.offset.Val)
+			}
+		}
+		if segment.offset.Val+segment.size.Val > imageMaxSize {
+			resizeSegment(segment, imageMaxSize-segment.offset.Val)
+		}
+		if segment.offset.Val+segment.size.Val > newSize {
+			newSize = segment.offset.Val + segment.size.Val
+		}
+	}
+	if newSize > imageMaxSize {
+		newSize = imageMaxSize
+	}
+	parsed.size.Val = newSize
+
+	// Drop 0-size segments.
+	parsed.filterSegments(func(i int, segment *mountImageSegment) bool {
+		return segment.size.Val > 0
+	})
+}
+
+type sortSegments struct {
+	parsed *mountImageArgs
+}
+
+func (s sortSegments) Len() int { return len(s.parsed.segments) }
+func (s sortSegments) Swap(i, j int) {
+	inner := s.parsed.segmentsGroup.Inner
+	inner[i], inner[j] = inner[j], inner[i]
+	s.parsed.segments[i], s.parsed.segments[j] = s.parsed.segments[j], s.parsed.segments[i]
+}
+func (s sortSegments) Less(i, j int) bool {
+	segments := s.parsed.segments
+	if segments[i].offset.Val != segments[j].offset.Val {
+		return segments[i].offset.Val < segments[j].offset.Val
+	}
+	return segments[i].size.Val < segments[j].size.Val
+}
+
+func (arch *arch) fixUpSyzMountImage(c *prog.Call) {
+	// Previously we did such a sanitization right in the common_linux.h, but this was problematic
+	// for two reasons:
+	// 1) It further complicates the already complicated executor code.
+	// 2) We'd need to duplicate the logic in Go for raw image extraction.
+	// So now we do all the initialization in Go and let the C code only interpret the commands.
+	ret, err := parseSyzMountImage(c)
+	if err != nil {
+		deactivateSyzMountImage(c)
+		return
+	}
+	const maxImageSegments = 4096
+	ret.filterSegments(func(i int, _ *mountImageSegment) bool {
+		return i < maxImageSegments
```

Comment:
This can be combined with the first filterSegments call inside of fixUpImageSegments.